modal/_utils/pattern_matcher.py

```diff
@@ -0,0 +1,274 @@
+# Copyright Modal Labs 2024
+"""Pattern matching library ported from https://github.com/moby/patternmatcher.
+
+This is the same pattern-matching logic used by Docker, except it is written in
+Python rather than Go. Also, the original Go library has a couple deprecated
+functions that we don't implement in this port.
+
+The main way to use this library is by constructing a `PatternMatcher` object,
+then asking it whether file paths match any of its patterns.
+"""
+
+import enum
+import os
+import re
+from typing import List, Optional, TextIO
+
+escape_chars = frozenset(".+()|{}$")
+
+
+class MatchType(enum.IntEnum):
+    UNKNOWN = 0
+    EXACT = 1
+    PREFIX = 2
+    SUFFIX = 3
+    REGEXP = 4
+
+
+class Pattern:
+    """Defines a single regex pattern used to filter file paths."""
+
+    def __init__(self) -> None:
+        """Initialize a new Pattern instance."""
+        self.match_type = MatchType.UNKNOWN
+        self.cleaned_pattern = ""
+        self.dirs: List[str] = []
+        self.regexp: Optional[re.Pattern] = None
+        self.exclusion = False
+
+    def __str__(self) -> str:
+        """Return the cleaned pattern as the string representation."""
+        return self.cleaned_pattern
+
+    def compile(self, separator: str) -> None:
+        """Compile the pattern into a regular expression.
+
+        Args:
+            separator (str): The path separator (e.g., '/' or '\\').
+
+        Raises:
+            ValueError: If the pattern is invalid.
+        """
+        reg_str = "^"
+        pattern = self.cleaned_pattern
+
+        esc_separator = separator
+        if separator == "\\":
+            esc_separator = "\\\\"
+
+        self.match_type = MatchType.EXACT
+        i = 0
+        pattern_length = len(pattern)
+        while i < pattern_length:
```

Comment:
This state machine is quite opaque to me, but the test coverage is good.

Reply:
Yeah, I think the important thing is that it is translated exactly from the upstream